airbyte-commons-server/src/main/java/io/airbyte/commons/server/handlers/helpers/OAuthSecretHelper.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.commons.server.handlers.helpers;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.common.annotations.VisibleForTesting;
+import io.airbyte.commons.json.Jsons;
+import io.airbyte.config.StandardSourceDefinition;
+import io.airbyte.protocol.models.ConnectorSpecification;
+import io.airbyte.validation.json.JsonValidationException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+/**
+ * Holds helpers to handle OAuth secrets.
+ */
+public class OAuthSecretHelper {
+
+  /**
+   * Should mimic frontend in using connector specification to figure out what auth properties are
+   * necessary for source creation and where to put them.
+   */
+  public static JsonNode setSecretsInConnectionConfiguration(final StandardSourceDefinition sourceDefinition,
+                                                             final JsonNode hydratedSecret,
+                                                             final JsonNode connectionConfiguration)
+      throws JsonValidationException {
+    final ConnectorSpecification spec = sourceDefinition.getSpec();
+
+    // Get the paths from advancedAuth that we need
+    final ObjectNode newConnectionConfiguration = connectionConfiguration.deepCopy();
+    if (spec.getAdvancedAuth() != null) {
+      final Map<String, List<String>> oAuthPaths = getOAuthConfigPaths(spec);
+      for (final Entry<String, List<String>> entry : oAuthPaths.entrySet()) {
+        // Key where we need to stuff things
+        final String key = entry.getKey();
+        final List<String> jsonPathList = entry.getValue();
+
+        Jsons.setNestedValue(newConnectionConfiguration, jsonPathList, hydratedSecret.get(key));
+      }
+      return newConnectionConfiguration;
+    } else {
+      // Just merge, complete_oauth handled setting rootNode for us.
+      return Jsons.mergeNodes(connectionConfiguration, hydratedSecret);
+    }
+  }
```

Comment:
@flash1293 a quick review here to make sure we're matching up with what frontend does would be awesome.

Reply:
Went through the code and from what I can tell it encodes the same logic like https://github.com/airbytehq/airbyte-cloud/blob/873c4ae061b1efda14f31265d50323b4be42b9a5/oss/airbyte-webapp/src/views/Connector/ConnectorForm/components/Sections/auth/useOauthFlowAdapter.tsx#L33

Did not test locally.